### DISCUSSION TITLE
Extend ways to implement nested partial classes

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#4993] [Language Ideas]
Currently we can only implement a partial class at the same nesting level:

```csharp
public class C
{
    public partial class D
    {
    }
    
    public partial class D // ok
    {
    }
}
```

There's no way for generators to generate an implementation outside the class `C` without marking `C` partial too.

I propose using `+` to imply nested classes and allow below code to be compiled:
```csharp
public class C
{
    public partial class D
    {
    }
}

public partial class C+D // ok
{
}
```

or use `.` instead of `+`:

```csharp
public class C
{
    public partial class D
    {
    }
}

public partial class C.D // ok
{
}
```

**Comment by YairHalberstadt** [comment]
Is the to make generating nested classes easier? I found it pretty simple to handle this in Stronginject.

**Reply by hez2010** [reply]
Yeah. It can also be extended for nested classes generation:
```csharp
class A+B+C+D { } // class A { class B { class C { class D { } } } }
// or
class A.B.C.D { } // class A { class B { class C { class D { } } } }
```

But I don't see the worthy of doing so, why would people generate a nested class directly without any code outside the nested one?

**Comment by svick** [comment]
How is this better than making the outer class `partial` also?

```c#
public partial class C
{
    public partial class D
    {
    }
}

public partial class C
{
    public partial class D // ok
    {
    }
}
```

**Reply by hez2010** [reply]
Then we have to make sure all classes on the nesting path to be partial before generators can generator code for an inside one.

**Reply by svick** [reply]
That doesn't seem like a big hurdle to me, especially since multiple levels of nesting tend to be quite rare.

**Reply by RikkiGibson** [reply]
It feels like it would be reasonable for a generator which requires this to give an error diagnostic here.

That said, it does seem odd to declare a partial type inside of a non-partial type. The utility of partials is really not to declare them multiple times within the same file, but rather multiple times across different files. So it could be reasonable to have a warning wave or a configurable analyzer or something to catch it in general.￼￼

**Reply by RikkiGibson** [reply]
Regarding the proposed design: it feels like it would only be worthwhile to introduce this if there was a strong use case where someone wanted to forbid adding additional members to some type, but they're OK with the user adding members to a nested partial type that already exists in the "main" declaration.￼
